Clarifications to SSO login/UIA (#2608)
matrix-org/matrix-doc: specification/modules/sso_login.rst

   which is delegating user authentication to the authentication server.

 * An "authentication server", which is responsible for authenticating the
   user.

This specification is concerned only with communication between the Matrix
client and the homeserver, and is independent of the SSO protocol used to
communicate with the authentication server. Different Matrix homeserver
implementations might support different SSO protocols.

Clients and homeservers implementing the SSO flow will need to consider both login_
and `user-interactive authentication`_. The flow is
similar in both cases, but there are slight differences.

Typically, SSO systems require a single "callback" URI to be configured at the
authentication server. Once the user is authenticated, their browser is
redirected to that URI. It is up to the Matrix homeserver implementation to
implement a suitable endpoint. For example, for CAS authentication the
homeserver should provide a means for the administrator to configure where the
CAS server is and the REST endpoints which consume the ticket.

Client login via SSO
---------------------

An overview of the process is as follows:

> 0. The Matrix client calls |GET /login|_ to find the supported login

Reviewer: Please can we start numbered lists at one?

Author: we *can*, though in this case step zero felt very much like a preliminary step.

Reviewer: It's just a bit awkward to have things start at zero, as it implies that it's not a real step (at least imo).

Author: that implication is correct, imo :)

   ``/cas/redirect`` and for ``m.login.sso`` use ``/sso/redirect`` (described below).
   The endpoints are otherwise the same.

Client behaviour
~~~~~~~~~~~~~~~~

The client starts the process by instructing the browser to navigate to
|/login/sso/redirect|_ with an appropriate ``redirectUrl``. Once authentication
is successful, the browser will be redirected to that ``redirectUrl``.

{{sso_login_redirect_cs_http_api}}

Security considerations
+++++++++++++++++++++++

1. CSRF attacks via manipulation of parameters on the ``redirectUrl``

   Clients should validate any requests to the ``redirectUrl``. In particular, it
   may be possible for attackers to falsify any query parameters, leading to
   cross-site request forgery (CSRF) attacks.

   For example, consider a web-based client at ``https://client.example.com``,
   which wants to initiate SSO login on the homeserver at ``server.example.org``.
   It does this by storing the homeserver name in a query parameter for the
   ``redirectUrl``: it redirects to
   ``https://server.example.org/login/sso/redirect?redirectUrl=https://client.example.com?hs=server.example.org``.

   An attacker could trick a victim into following a link to

>    ``https://server.example.org/login/sso/redirect?redirectUrl=https://client.example.com?hs=evil.org``,

Author: ```suggestion
   ``https://server.example.org/login/sso/redirect?redirectUrl=https://client.example.com?hs=evil.com``,
```

   which would result in the client sending a login token for the victim's

>    account to the attacker-controlled site ``evil.com``.

Reviewer: example was .org
```suggestion
   account to the attacker-controlled site ``evil.org``.
```

Author: yeah. will change the example instead for consistency.

   endpoint, and to give more meaningful errors in the case of
   faulty/poorly-configured clients.

Handling the callback from the Authentication server
++++++++++++++++++++++++++++++++++++++++++++++++++++

Note that there will normally be a single callback URI which is used for both login
and user-interactive authentication: it is up to the homeserver implementation
to distinguish which is taking place.

The homeserver should validate the response from the SSO system: this may
require additional calls to the authentication server, and/or may require
checking a signature on the response.

The homeserver then proceeds as follows:

#. The homeserver MUST map the user details received from the authentication

>    server to a valid `Matrix user identifier <../index.html#user-identifiers>`_.

Reviewer: ```suggestion
   server to a valid `Matrix user identifier <../appendices.html#user-identifiers>`_.
```

   The guidance in `Mapping from other character sets

>    <../index.html#mapping-from-other-character-sets>`_ may be useful.

Reviewer: ```suggestion
   <../appendices.html#mapping-from-other-character-sets>`_ may be useful.
```

#. If the generated user identifier represents a new user, it should be
   registered as a new user.

#. The homeserver should generate a short-term login token. This is an opaque
   token, suitable for use with the ``m.login.token`` type of the |/login|_
   API. The lifetime of this token SHOULD be limited to around five

>    seconds.

Reviewer 2: This seems to assume relatively quick roundtrip between the client and the homeserver?

#. The homeserver adds a query parameter of ``loginToken``, with the value of
   the generated login token, to the ``redirectUrl`` given in the
   ``/_matrix/client/%CLIENT_MAJOR_VERSION%/login/sso/redirect``
   request. (Note: ``redirectURL`` may or may not include existing query
   parameters. If it already includes one or more ``loginToken`` parameters,
   they should be removed before adding the new one.)

> #. The homeserver redirects the user's browser to the URI thus built.

Reviewer 2: ```suggestion
#. The homeserver redirects the client to the URI thus built.
```

Security considerations
~~~~~~~~~~~~~~~~~~~~~~~

1. Homeservers should ensure that login tokens are not sent to malicious
   clients.

   For example, consider a homeserver at ``server.example.org``. An attacker tricks
   a victim into following a link to
   ``https://server.example.org/login/sso/redirect?redirectUrl=https://evil.com``,
   resulting in a login token being sent to the attacker-controlled site

>    ``evil.com``. This is a form of cross-site request forgery (CSRF).

Reviewer: Ah, I see where the .com came from now - we should probably be consistent in the examples.

See also the "Security considerations" below.

Handling the callback from the Authentication server
++++++++++++++++++++++++++++++++++++++++++++++++++++

Note that there will normally be a single callback URI which is used for both login
and user-interactive authentication: it is up to the homeserver implementation
to distinguish which is taking place.

The homeserver should validate the response from the SSO system: this may
require additional calls to the authentication server, and/or may require
checking a signature on the response.

The homeserver then returns the `user-interactive authentication fallback
completion`_ page to the user's browser.

Security considerations
+++++++++++++++++++++++

1. Confirming the operation

   The homeserver SHOULD confirm that the user is happy for the operation to go
   ahead. The goal of the user-interactive authentication operation is to guard
   against a compromised ``access_token`` being used to take over the user's
   account. Simply redirecting the user to the SSO system is insufficient,
   since they may not realise what is being asked of them, or the SSO system
   may even confirm the authentication automatically.

>    For example, the homeserver might serve a page with words to the effact of:

Reviewer: ```suggestion
   For example, the homeserver might serve a page with words to the effect of:
```

     A client is trying to remove a device from your account. To confirm this
     action, re-authenticate with single sign-on. If you did not expect this, your
     account may be compromised!

   This confirmation could take place before redirecting to the SSO
   authentication page (when handling the

>    ``/_matrix/client/%CLIENT_MAJOR_VERSION%/auth/m.login.sso/fallback/web``

Reviewer: fwiw we have a history of shortening URLs to after the version (ie: `/auth/m.login.sso/fallback/web`), largely for legibility.

Author: mmhmm. There was some inconsistency here and I went this way: do you think I should go the other?

Reviewer: I don't personally have a preference, just mentioning it in the event it felt annoying to have to do it every time.
